### PR TITLE
Version support

### PR DESCRIPTION
**Pull request by spangenberg**
Introduce `version` support for user agents and CLI retrieval trough `-version` flag.
Version gets set during building process of AWS actuator.

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -25,12 +26,18 @@ import (
...
 	awsclient "sigs.k8s.io/cluster-api-provider-aws/pkg/client"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/version"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
 )
 
 func main() {
+	if len(os.Args) > 1 && os.Args[1] == "version" {
+		fmt.Println(version.String)
```
Comment (paulfantom) [inline cmd/manager/main.go]: why not `glog.fatal(...)` instead of fmt?

Comment (spangenberg) [reply cmd/manager/main.go]: Which one do you prefer if you execute `manager version`, think this is more a command for humans.

`ClusterAPIProviderAWS v0.2.0-64-g0afb3b90-dirty`

`F0315 15:14:05.721565   95782 main.go:48] ClusterAPIProviderAWS v0.2.0-64-g0afb3b90-dirty`

Comment (bison) [reply cmd/manager/main.go]: Strong +1 for just using `fmt.Println()` here.

File: Makefile [hunk 2]
```diff
@@ -58,7 +58,7 @@ bin:
 
 .PHONY: build
 build: ## build binaries
-	$(DOCKER_CMD) go build $(GOGCFLAGS) -o bin/machine-controller-manager -ldflags '-extldflags "-static"' sigs.k8s.io/cluster-api-provider-aws/cmd/manager
+	$(DOCKER_CMD) go build $(GOGCFLAGS) -o bin/machine-controller-manager -ldflags '-X sigs.k8s.io/cluster-api-provider-aws/pkg/version.Raw=$(shell git describe --abbrev=8 --dirty --always) -extldflags "-static"' sigs.k8s.io/cluster-api-provider-aws/cmd/manager
```
Comment (bison) [inline Makefile]: Wouldn't block the PR on this, but this is a little hard to read. Elsewhere we do this at the top of the Makefile:
```makefile
VERSION     ?= $(shell git describe --always --dirty --abbrev=7)
LD_FLAGS    ?= -X $(REPO_PATH)/pkg/version.Raw=$(VERSION)
BUILD_DEST  ?= bin/cluster-autoscaler-operator
```

Then in the build target:
```makefile
.PHONY: build
build: ## build binaries
        $(DOCKER_CMD) go build $(GOGCFLAGS) -ldflags "$(LD_FLAGS)" -o "$(BUILD_DEST)" "$(REPO_PATH)/cmd/manager"
```

Comment (spangenberg) [reply Makefile]: Thanks, will change it.

File: cmd/manager/main.go (hunk 1 above)
Comment (paulfantom) [reply cmd/manager/main.go]: Makes sense :+1:

**Comment by ingvagabund** [comment]
```
$ make build
go build  -o "bin/machine-controller-manager" -ldflags "-X sigs.k8s.io/cluster-api-provider-aws/pkg/version.Raw=v0.2.0-64-gcd8ea89 -extldflags "-static"" "sigs.k8s.io/cluster-api-provider-aws/cmd/manager"
go build  -o bin/manager -ldflags '-extldflags "-static"' sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/openshift/cluster-api/cmd/manager
$ ./bin/machine-controller-manager version
ClusterAPIProviderAWS v0.2.0-64-gcd8ea89
```

**Comment by ingvagabund** [comment]
```
$ make images
...
Complete!
Loaded plugins: fastestmirror, ovl
Cleaning repos: base cbs-paas7-openshift-multiarch-el7-build extras updates
Cleaning up list of fastest mirrors
--> COPY --from=builder /go/src/sigs.k8s.io/cluster-api-provider-aws/manager /
--> COPY --from=builder /go/src/sigs.k8s.io/cluster-api-provider-aws/machine-controller-manager /
--> Committing changes to origin-aws-machine-controllers:v0.2.0-64-gcd8ea89 ...
--> Tagged as origin-aws-machine-controllers:latest
--> Done
$ docker run -it origin-aws-machine-controllers:v0.2.0-64-gcd8ea89 /machine-controller-manager version
ClusterAPIProviderAWS v0.0.0-was-not-built-properly
```

@spangenberg the Dockerfile needs to be updated as well. Feel free to check @paulfantom 's changes in mao.

File: Dockerfile [hunk 3]
```diff
@@ -1,8 +1,7 @@
 FROM registry.svc.ci.openshift.org/openshift/release:golang-1.10 AS builder
 WORKDIR /go/src/sigs.k8s.io/cluster-api-provider-aws
 COPY . .
-RUN go build -o ./machine-controller-manager ./cmd/manager
-RUN go build -o ./manager ./vendor/github.com/openshift/cluster-api/cmd/manager
+RUN NO_DOCKER=1 make build
```
Comment (enxebre) [inline Dockerfile]: how is this related to this PR/commit?

Comment (ingvagabund) [reply Dockerfile]: https://github.com/openshift/cluster-api-provider-aws/pull/179#issuecomment-473830832

Comment (spangenberg) [reply Dockerfile]: Thanks @ingvagabund, yeah it basically removes duplicate logic by just running the build trough make for the image release process.

File: cmd/manager/main.go [hunk 4]
```diff
@@ -26,12 +27,18 @@ import (
...
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1beta1"
 	awsclient "sigs.k8s.io/cluster-api-provider-aws/pkg/client"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/version"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
 )
 
 func main() {
+	if len(os.Args) > 1 && os.Args[1] == "version" {
```
Comment (bison) [inline cmd/manager/main.go]: We're already parsing the flags in order to initialize klog. It seems preferable to just use a flag for this:
```golang
var printVersion bool

func init() {
  flag.BoolVar(&printVersion, "version", false, "print version and exit")
}

func main() {
  ...
  flag.Parse()

  if printVersion {
    fmt.Println(version.String)
    os.Exit(0)
  }
}
```

**Comment by ingvagabund** [comment]
```
$ docker run -it origin-aws-machine-controllers:v0.2.0-78-gafafd8a /machine-controller-manager version
F0415 12:25:29.030293       1 main.go:60] Error getting configuration: could not locate a kubeconfig
goroutine 1 [running]:
sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.stacks(0xc420447000, 0xc420476000, 0x65, 0xba)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:769 +0xcf
sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).output(0x22e79e0, 0xc400000003, 0xc420472210, 0x2269d75, 0x7, 0x3c, 0x0)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:720 +0x32d
sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).printf(0x22e79e0, 0xc400000003, 0x16b24c3, 0x1f, 0xc4204adeb8, 0x1, 0x1)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:655 +0x14b
sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.Fatalf(0x16b24c3, 0x1f, 0xc4204adeb8, 0x1, 0x1)
	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:1148 +0x67
main.main()
	/go/src/sigs.k8s.io/cluster-api-provider-aws/cmd/manager/main.go:60 +0x180
```

**Comment by spangenberg** [comment]
> ```
> $ docker run -it origin-aws-machine-controllers:v0.2.0-78-gafafd8a /machine-controller-manager version
> F0415 12:25:29.030293       1 main.go:60] Error getting configuration: could not locate a kubeconfig
> goroutine 1 [running]:
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.stacks(0xc420447000, 0xc420476000, 0x65, 0xba)
> 	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:769 +0xcf
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).output(0x22e79e0, 0xc400000003, 0xc420472210, 0x2269d75, 0x7, 0x3c, 0x0)
> 	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:720 +0x32d
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).printf(0x22e79e0, 0xc400000003, 0x16b24c3, 0x1f, 0xc4204adeb8, 0x1, 0x1)
> 	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:655 +0x14b
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.Fatalf(0x16b24c3, 0x1f, 0xc4204adeb8, 0x1, 0x1)
> 	/go/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:1148 +0x67
> main.main()
> 	/go/src/sigs.k8s.io/cluster-api-provider-aws/cmd/manager/main.go:60 +0x180
> ```
> ```
> $ ./bin/machine-controller-manager version
> F0415 14:26:34.037784   13871 main.go:83] Error adding actuator: no matches for kind "Machine" in version "machine.openshift.io/v1beta1"
> goroutine 1 [running]:
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.stacks(0xc000200500, 0xc000504540, 0x89, 0xde)
> 	/home/jchaloup/Projects/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:769 +0xd4
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).output(0x278cc80, 0xc000000003, 0xc00013c0b0, 0x2709809, 0x7, 0x53, 0x0)
> 	/home/jchaloup/Projects/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:720 +0x329
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.(*loggingT).printf(0x278cc80, 0x3, 0x17031a5, 0x19, 0xc0007dfe88, 0x1, 0x1)
> 	/home/jchaloup/Projects/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:655 +0x14b
> sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog.Fatalf(0x17031a5, 0x19, 0xc0007dfe88, 0x1, 0x1)
> 	/home/jchaloup/Projects/src/sigs.k8s.io/cluster-api-provider-aws/vendor/github.com/golang/glog/glog.go:1148 +0x67
> main.main()
> 	/home/jchaloup/Projects/src/sigs.k8s.io/cluster-api-provider-aws/cmd/manager/main.go:83 +0x456
> ```

It's added as a flag, try `-version`.

**Comment by ingvagabund** [comment]
```
$ ./bin/machine-controller-manager --version
ClusterAPIProviderAWS v0.2.0-78-gafafd8a
$ docker run -it origin-aws-machine-controllers:v0.2.0-78-gafafd8a /machine-controller-manager --version
ClusterAPIProviderAWS 1.10.3
```

**Comment by ingvagabund** [comment]
> It's added as a flag, try -version.
Yeah, just noticed. Can you update the PR description accordingly?

**Comment by spangenberg** [comment]
> > It's added as a flag, try -version.
> > Yeah, just noticed. Can you update the PR description accordingly?

Sure, thanks for pointing that out.

**Comment by ingvagabund** [comment]
@spangenberg can you share you log from `make images`?

**Comment by spangenberg** [comment]
> @spangenberg can you share you log from `make images`?

Thanks for pointing out the version miss-match, it's building now properly:
```sh
go build  -o "bin/machine-controller-manager" \
               -ldflags "-X sigs.k8s.io/cluster-api-provider-aws/pkg/version.Raw=v0.2.0-78-gafafd8a -extldflags "-static"" "sigs.k8s.io/cluster-api-provider-aws/cmd/manager"
```
The go version took precedence in the build.

**Comment by ingvagabund** [comment]
```
$ docker run -it origin-aws-machine-controllers:v0.2.0-78-gb24f19d /machine-controller-manager --version
ClusterAPIProviderAWS v0.2.0-78-gb24f19d
$ ./bin/machine-controller-manager --version
ClusterAPIProviderAWS v0.2.0-78-gb24f19d
[jchaloup@dhcp-24-170 cluster-api-provider-aws]$ 
```

looks good now.

/approve
